Fixes #10248 - cloning volume information for virtual machines

tstrachota opened the pull request:
`vm.attributes` (Fog) returns volumes and nics info as objects under keys `:volumes` and `:interfaces`. This PR places the volume attributes into `:volumes_attributes` where the Rails' form helper can find it.

NICs are no longer a problem since we save the original compute attributes with them now.

Tested providers checklist:
- [x] libvirt
- [x] oVirt
- [x] VMWare

Commit:
Fixes #10248 - cloning volume information for virtual machines

domcleal commented:
Should this be in all compute resources that have volumes?  Can it move into compute_resource.rb?

tstrachota commented:
I didn't have a chance to test it elsewhere but it probably makes sense. Let me test it for libvirt and I'll let you know.

tstrachota commented:
Ok, libvirt has the same issue and the fix it seems to be working for it too. I made the code more defensive and moved it to compute_resource.rb. I also fixed one more issue with memory count for libvirt.
We need to test it with oVirt prior to merge.

domcleal commented:
:+1: on the direction, this looks much better.  As you say, needs some testing on each CR to check coverage.

dLobatog commented:
Can you write a test for this one? I'm not sure exactly what this is doing and I think someone could break it in the future.

domcleal commented:
Also the commit message needs rewriting.

domcleal commented:
Works very well on libvirt.

@shlomizadok, could you test behaviour on vSphere please?  Check that all storage volume related attrs, like location, size etc are copied accurately.

ditto anybody with access to oVirt.  @lzap?

tstrachota commented:
Commit message updated. I still need to have a look at the tests.

shlomizadok commented:
ACK for vmware

dLobatog commented on app/models/compute_resource.rb:
```diff
@@ -244,7 +244,12 @@ def compute_profile_attributes_for(id)
   def vm_compute_attributes_for(uuid)
     vm = find_vm_by_uuid(uuid)
-    vm.attributes.reject{|k,v| k == :id }
+    vm_attrs = vm.attributes.reject{|k,v| k == :id }
```
Cloning AWS hosts is currently failing with `undefined method`attributes' for nil:NilClass` , at least if the Host is associated with an EC2 virtual machine that doesn't exist any more.

dLobatog commented:
Seems to work well on GCE even when the associated virtual machine doesn't exist anymore. On libvirt I'm getting this error:

```
no implicit conversion of nil into String
app/models/compute_resources/foreman/model/libvirt.rb:29:in `find_vm_by_uuid'
app/models/compute_resource.rb:246:in `vm_compute_attributes_for'
app/models/compute_resources/foreman/model/libvirt.rb:155:in `vm_compute_attributes_for'
app/models/host/managed.rb:627:in `clone'
app/controllers/hosts_controller.rb:74:in `clone'
```

@tstrachota I think it'd be really useful to try to cover up this situation and allow people to clone the host and create a new vm with the same settings, especially in this case where the vm doesn't exist any more in the compute resource.

tstrachota commented:
@eLobato good point! I added handling of situations when the vm can't be found and also unified behaviour across compute profiles in such case. Some of them used to just return `nil` instead of raising `RecordNotFound`.

It's in a separate commit for easier review. It should be sqashed prior to merge.

Commit:
Refs #10248 - handling a state when a vm can't be found

dLobatog commented:
I think I'm still getting the same `no implicit conversion of nil into String` with Libvirt and a non-existent host unfortunately. I like the unification, it's a shame we can't just rescue `Fog::Errors::Error` instead of individual CR errors - maybe look if all the CR errors inherit from that but I doubt it.

Tests seem repetitive, perhaps you can define a method for this test and call it from the other unit tests with different arguments.

lzap commented:
All my ovirts are dead, but planning new setup. Does it still need testing with oVirt?

@tstrachota can you edit your description and provide a check list with all the CR that has been tested already?

domcleal commented:
> Does it still need testing with oVirt?

Yes, it will when updated.

tstrachota commented:
@eLobato I can't reproduce the `implicit conversion...` issue. I create a host on libvirt, delete it via virt-manager and try to clone from host index page. It works for me. Can you please help me with the reproducer? Is there anything you do differently?

tstrachota commented:
Tests refactored.

domcleal commented:
@tstrachota they're red

Commit:
Refs #10248 - refactoring the tests

tstrachota commented:
Rebased

dLobatog commented:
@tstrachota Not sure what was it but I can clone stuff on Libvirt just fine now, no `implicit conversion` error whatsoever. Still someone should test on oVirt, other than that it works. I'll check the code now.

dLobatog commented on app/models/compute_resource.rb:
```diff
@@ -244,7 +244,16 @@ def compute_profile_attributes_for(id)
+      vm_attrs[:volumes_attributes] = Hash[volumes.each_with_index.map { |volume, idx| [idx.to_s, volume.attributes] }]
+    end
+    vm_attrs
+  rescue ActiveRecord::RecordNotFound
```
What could throw this in this method? It seems that these are calls to the compute resource.

tstrachota replied:
That's the unification of `find_vm_by_uuid(uuid)` exceptions. When the vm is not found `RecordNotFound` is raised.

dLobatog commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -150,6 +150,12 @@ def associated_host(vm)
 
+    def vm_compute_attributes_for(uuid)
+      vm_attrs = super
+      vm_attrs[:memory] = vm_attrs[:memory_size]*1024 rescue nil # value is returned in megabytes, we need bytes
```
rescue to get the error and backtrace printed in debug at least, that could be useful when debugging problems with this area of the code.

dLobatog commented on test/unit/compute_resources/cr_test_helpers.rb:
```diff
@@ -0,0 +1,27 @@
+module CRTestHelpers
```
Small quibble, `ComputeResourceTestHelper` seems more idiomatic & friendly for newcomers who might not know what a `CR` is.

dLobatog commented:
Overall :+1: , just left a couple of questions inline, thanks @tstrachota !

domcleal commented:
I'd like to ensure we can merge #2492 after this PR in some form, which is using the same method to output the attributes hash over the API.  I have a couple of specific questions:
1. Here, `:volumes` is called `:volumes_attributes` - should we rename that and add `_attributes` in `Host::Managed#clone` or remove the _attributes suffix in `Host::Managed#vm_compute_attributes`?
2. #2492 adds interfaces too - will that case any issues with cloning, or is it even beneficial?  Should we add it through that PR or ignore it during cloning?

tstrachota commented:
@domcleal https://github.com/theforeman/foreman/pull/2492 does more or less the same but produces output that can't be re-used as a new input for vm create (in cloning). I skipped network interfaces intentionally. We handle them in the `Nic` model now. I can see 2 ways out of it in https://github.com/theforeman/foreman/pull/2492:
1. Don't include network interfaces in the output of the API call and keep `volumes_attributes` key.
   Currently in all UI, API and CLI the compute attributes for nics are passed separately in `host[interfaces_attributes][compute_attributes]` so the API output would be consistent with expected input. The rabl for `interfaces#show` should be extended to return compute attributes too.
2. Include both volumes and interfaces in the output of the API call but collect the information outside the `vm_attributes_for` method which should return data in the expected format. The pro here is that the API would return complete information. On the other hand it could be misleading that what is returned can't be used as input.

I prefer No 1 as it's closer to how people use the API from my experience.

Commit:
Refs #10248 - test helper renamed & some debug logging

tstrachota commented:
@eLobato updated according to your comments.

dLobatog commented:
I'll try to test this with oVirt tomorrow if I can get some cloud instance to install it, seems ready in all other CRs.

tstrachota commented:
@eLobato @domcleal I rebased the PR and fixed some issues that arose after that. I also tested oVirt and it seems to work fine. Let me know if you want to re-check it personally. I'll send you details about the oVirt instance.

tstrachota commented:
[test]

domcleal commented on app/helpers/hosts_and_hostgroups_helper.rb:
```diff
@@ -60,7 +60,7 @@ def puppet_ca(f, can_override, override)
              { :include_blank => blank_or_inherit_f(f, :puppet_ca_proxy),
                :disable_button => can_override ? _(INHERIT_TEXT) : nil,
-               :disable_button_enabled => override && !params[:host][:puppet_ca_proxy_id],
+               :disable_button_enabled => override && !(params[:host] && params[:host][:puppet_ca_proxy_id]),
```
This looks like a fix for #11255, which already has a PR open - could you remove it?

domcleal commented on app/models/compute_resource.rb:
```diff
@@ -244,7 +244,17 @@ def compute_profile_attributes_for(id)
+    end
+    vm_attrs
+  rescue ActiveRecord::RecordNotFound
+    logger.debug("vm with uuid '#{uuid}' not found")
```
I'd consider this info or warn even, and consider adding the CR (self) to the log message.

domcleal commented on app/views/hosts/_form.html.erb:
```diff
@@ -60,7 +60,7 @@
           <%= select_f f, :compute_profile_id, ComputeProfile.visibles, :id, :name,
             { :disable_button => _(HostsAndHostgroupsHelper::INHERIT_TEXT),
-              :disable_button_enabled => @host.hostgroup && @host.hostgroup_id_was.nil? && !params[:host][:compute_profile_id],
+              :disable_button_enabled => @host.hostgroup && @host.hostgroup_id_was.nil? && !(params[:host] && params[:host][:compute_profile_id]),
```
ditto about other PR

domcleal commented on test/unit/host_test.rb:
```diff
@@ -148,7 +148,7 @@ class HostTest < ActiveSupport::TestCase
 
   test "can fetch vm compute attributes" do
-    host = FactoryGirl.create(:host, :compute_resource => compute_resources(:one))
+    host = FactoryGirl.create(:host, :compute_resource => compute_resources(:ec2))
```
Any idea why that was?

tstrachota replied:
`:one` is libvirt which handles the compute parameters a bit differently so I switched it to `:ec2`. I'll add unit tests for the other compute resources too.

domcleal commented on app/assets/javascripts/compute_resources/ovirt/nic_info.js:
```diff
@@ -1,3 +1,3 @@
 providerSpecificNICInfo = function(form) {
-  return form.find('.ovirt_name').val() + ' @ ' + form.find('.ovirt_network').val();
+  return form.find('.ovirt_name').val() + ' @ ' + form.find('.ovirt_network').text();
```
Do you know what exactly that fixes?  Is it a bug with something like select2?  (If so, please file it separately as we'd take it for 1.9.)

tstrachota replied:
It fixes my own mistake. Value in the selectbox is some hash, while the text is name of the network. So without this change the js used to produce something like `eth0 @ b3499d8ae083` now it's `eth0 @ local_network`

domcleal commented:
Thanks, that's good enough for me.  The fixes after rebases all look like they belong elsewhere.

Otherwise it looks OK.  Any chance you can add unit tests for the other CRs (oVirt, OpenStack), if you know their behaviour?

tstrachota commented:
Fixed the comments and added tests for finding vms on OpenStack and oVirt.

Commit:
Refs #10248 - fixes NIC label for oVirt

Commit:
Fixes #10248 - more tests for finding vm by uuid

- unit tests for find_vm_by_uuid on oVirt and OpenStack
- unification for vm not found exception on oVirt
- updated warning when the vm is not found

domcleal commented:
Merged as 4269abb, thanks @tstrachota!
